lib/spack/spack/test/conftest.py

```diff
@@ -270,31 +270,74 @@ def _skip_if_missing_executables(request):
 spack.architecture.real_platform = spack.architecture.platform
 spack.architecture.platform = lambda: spack.platforms.test.Test()
 
-##########
-# Test-specific fixtures
-##########
+
+#
+# Context managers used by fixtures
+#
+# Because these context managers modify global state, they should really
+# ONLY be used persistently (i.e., around yield statements) in
+# function-scoped fixtures, OR in autouse session- or module-scoped
+# fixtures.
+#
+# If they're used regular tests or in module-scoped fixtures that are
+# then injected as funciton arguments, weird things can happen, because
+# the original state won't be restored until *after* the fixture is
+# destroyed.  This makes sense for an autouse fixture, where you know
+# everything in the module/session is going to need the modified
+# behavior, but modifying global state for one function in a way that
+# won't be restored until after the module or session is done essentially
+# leaves garbage behind for other tests.
+#
+# In general, we should module- or session-scope the *STATE* required for
+# these global objects, but we shouldn't module-or session scope their
+# *USE*, or things can get really confusing.
+#
+
+@contextlib.contextmanager
+def use_configuration(config):
+    """Context manager to swap out the global Spack configuration."""
+    saved = spack.config.config
+    spack.config.config = config
+    yield
+    spack.config.config = saved
 
 
+@contextlib.contextmanager
+def use_store(store):
+    """Context manager to swap out the global Spack store."""
+    saved = spack.store.store
+    spack.store.store = store
+    yield
+    spack.store.store = saved
+
+
+@contextlib.contextmanager
+def use_repo(repo):
+    """Context manager to swap out the global Spack repo path."""
+    with spack.repo.swap(repo):
+        yield
+
+#
+# Test-specific fixtures
+#
 @pytest.fixture(scope='session')
-def repo_path():
-    """Session scoped RepoPath object pointing to the mock repository"""
-    return spack.repo.RepoPath(spack.paths.mock_packages_path)
+def mock_repo_path():
+    yield spack.repo.RepoPath(spack.paths.mock_packages_path)
 
 
-@pytest.fixture(scope='module')
-def mock_packages(repo_path):
+@pytest.fixture(scope='function')
+def mock_packages(mock_repo_path):
     """Use the 'builtin.mock' repository instead of 'builtin'"""
-    mock_repo = copy.deepcopy(repo_path)
-    with spack.repo.swap(mock_repo):
-        yield
+    with use_repo(mock_repo_path):
+        yield mock_repo_path
```

Comment:
Minor: I think there might be an inconsistency between the name of the fixture `mock_packages` and the return value which is the path to the repository.

Reply:
I didn't change this one because `mock_packages` is used *everywhere*.  I suppose I can change this later if we want to be consistent (I kind of agree) but I think everyone just understands what `mock_packages` means at this point.